R/derive_param_extreme_event.R

```diff
@@ -0,0 +1,379 @@
+#'   filter_source = PARAMCD == "OVR" & AVALC == "PD",
+#'   order = vars(ADT),
+#'   new_var = AVALC,
+#'   true_value = "Y",
+#'   false_value = "N",
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     PARAMCD = "PD",
+#'     PARAM = "Disease Progression",
+#'     ANL01FL = "Y",
+#'     ADT = ADT
+#'   )
+#' )
+#'
+#' # derive parameter indicating death
+#' derive_param_extreme_event(
+#'   dataset = adrs,
+#'   dataset_adsl = adsl,
+#'   dataset_source = adsl,
+#'   filter_source = !is.na(DTHDT),
+#'   order = vars(DTHDT),
+#'   new_var = AVALC,
+#'   true_value = "Y",
+#'   false_value = "N",
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     PARAMCD = "DEATH",
+#'     PARAM = "Death",
+#'     ANL01FL = "Y",
+#'     ADT = DTHDT
+#'   )
+#' )
+derive_param_extreme_event <- function(dataset,
+                                       dataset_adsl,
+                                       dataset_source,
+                                       filter_source,
+                                       order,
+                                       new_var = AVALC,
+                                       true_value = "Y",
+                                       false_value = "N",
+                                       mode = "first",
+                                       subject_keys = vars(STUDYID, USUBJID),
+                                       set_values_to,
+                                       check_type = "warning") {
+  # Check input parameters
+  filter_source <- assert_filter_cond(enquo(filter_source))
+  assert_vars(order)
+  new_var <- assert_symbol(enquo(new_var))
+  assert_same_type(true_value, false_value)
+  assert_vars(subject_keys)
+  assert_data_frame(dataset, required_vars = vars(PARAMCD))
+  assert_data_frame(dataset_source, required_vars = vars(!!!subject_keys, !!!order))
+  assert_data_frame(dataset_adsl, required_vars = subject_keys)
+  check_type <-
+    assert_character_scalar(
+      check_type,
+      values = c("none", "warning", "error"),
+      case_sensitive = FALSE
+    )
+  mode <- assert_character_scalar(
+    mode,
+    values = c("first", "last"),
+    case_sensitive = FALSE
+  )
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+
+  # Create new observations
+  source_vars <- colnames(dataset_source)
+  adsl_vars <- colnames(dataset_adsl)
+
+  events <- dataset_source %>%
+    filter_if(filter_source) %>%
+    filter_extreme(
+      by_vars = subject_keys,
+      order = order,
+      mode = mode,
+      check_type = check_type
+    )
+
+  noevents <- anti_join(
+    select(dataset_adsl, intersect(source_vars, adsl_vars)),
+    select(events, !!!subject_keys)
+  )
+
+  new_obs <- bind_rows(yes = events, no = noevents, .id = "HAD_EVENT") %>%
+    mutate(
+      !!new_var := if_else(.data$HAD_EVENT == "yes", true_value, false_value),
+      !!!set_values_to
+    ) %>%
+    select(-.data$HAD_EVENT)
```

Comment:
Please do not use `.data$HAD_EVENT`. This was deprecated in the latest version of tidyselect (see https://tidyselect.r-lib.org/news/index.html#lifecycle-changes-1-2-0).
For temporary variables like `HAD_EVENT` you should use `get_new_tmp_var()`.
Maybe it is easier to add
```
  %>% mutate(!!new_var := true_value)
```
and
```
  %>% mutate(!!new_var := false_value)
```
to the creation of `events` and `noevents` above.

Reply:
Thanks for the heads up on temporary variables - I'll bear that in mind for future reference. Have implemented your suggestion `%>% mutate()`